src/openms/source/QC/Contaminants.cpp

```diff
@@ -35,113 +35,181 @@
 #include <OpenMS/QC/Contaminants.h>
 #include <include/OpenMS/CHEMISTRY/ProteaseDigestion.h>
 #include <include/OpenMS/METADATA/ProteinIdentification.h>
-#include <include/OpenMS/FORMAT/FeatureXMLFile.h>
-//#include <include/OpenMS/CHEMISTRY/DigestionEnzymeDB.h>
-//#include <include/OpenMS/CHEMISTRY/ProteaseDB.h>
 
-using namespace OpenMS;
-using namespace std;
 
+using namespace std;
 
 
-void Contaminants::compute(FeatureMap& features, const std::vector<FASTAFile::FASTAEntry>& contaminants)
+namespace OpenMS
 {
-  if (digested_db_.empty())
+
+  void Contaminants::compute(FeatureMap &features, const std::vector<FASTAFile::FASTAEntry> &contaminants)
   {
-    String enzyme = features.getProteinIdentifications()[0].getSearchParameters().digestion_enzyme.getName();
-    if (enzyme == "unknown_enzyme")
+    //empty FeatureMap
+    if (features.empty())
     {
-      LOG_WARN << "No digestion enzyme in featureMap detected.";
+      LOG_WARN << "FeatureMap is empty." << "\n";
+      return;
     }
-    ProteaseDigestion digestor;
-    digestor.setEnzyme(enzyme);
-    UInt missed_cleavages (features.getProteinIdentifications()[0].getSearchParameters().missed_cleavages);
-    digestor.setMissedCleavages(missed_cleavages);
-
-    Size dropped_by_length(0); // stats for removing candidates
-
-
-    for (const FASTAFile::FASTAEntry& fe : contaminants)
+    //empty features
+    if (features[0].getPeptideIdentifications().empty())
+    {
+      LOG_WARN << "Features are empty." << "\n";
+      return;
+    }
+    //empty contaminants database
+    if (contaminants.empty())
+    {
+      LOG_WARN << "FASTAFile is empty." << "\n";
+      return;
+    }
+    //fill the unordered set once with the digested contaminants database
+    if (digested_db_.empty())
     {
+      ProteaseDigestion digestor;
+      String enzyme = features.getProteinIdentifications()[0].getSearchParameters().digestion_enzyme.getName();
 
-      vector<AASequence> current_digest;
-      if (enzyme == "unknown_enzyme")                                                   //no enzyme given
+      //no enzyme is given
+      if (enzyme == "unknown_enzyme")
       {
-        current_digest.push_back(AASequence::fromString(fe.sequence));
+        LOG_WARN << "No digestion enzyme in featureMap detected." << "\n";
```

Comment:
replace with
`      throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No digestion enzyme in FeatureMap detected. No computation possible.");`

Reply:
maybe that was intentional and the contaminant database is already digested

if so: the user still gets his results 
if not: the user will get 0 contaminants and the warning that the database was not digested

should we still take the option of giving an already digested database from the user?

Reply:
exception is fine here.
The user can control the behavior by setting a digestion enzyme (also "none" is a valid option as enzyme -- the actual name might be diferent).